### ISSUE TITLE
Use a SUID/sudo-approved shim to create namespaces

### ISSUE THREAD

**Issue by LukeShu** [sandstorm-io/sandstorm#1664]
As noted in the install script, some distros (such as Arch Linux, or in my case, Parabola) have their kernels compiled with `CONFIG_USER_NS=n`.  This is because it has had, and likely still does, have multiple privilege-escalation security vulnerabilities.  Some distros patch the kernel to enable the feature at runtime via a sysctl flag.  On these systems, the Sandstorm installer asks the user to enable it globally.

In the case of Sandstorm, the vulnerabilities are mitigated by seccomp.  Unfortunately, not all of the other services on a system can be counted on to do that.  That is: Even if my kernel is patched to have the sysctl flag, turning it on globally scares the shit out of me.

This can be worked around by providing a SUID executable, or executable that gets configured in `/etc/sudoers` to run password-les; and using that to run things in the namespaces.  Besides supporting kernels with `CONFIG_USER_NS=n`, it would provide more granular control over which users can create namespaces.

**Comment by kentonv** [comment]
This has a number of problems, though:

* It would be impossible to install and run Sandstorm without root privileges if the supervisor had to be setuid-root. Therefore, we would need to keep the existing code path, which means we need to maintain and test two code paths that may have subtly different behavior.
* If user namespaces cannot be created even by root, then we have no way to force the UID inside the app container to 1000, which is likely to cause odd bugs in apps. For example, we've seen cases where programs crash if their UID is not in /etc/passwd, which is a file provided by the app.
* We currently run the entire server with NO_NEW_PRIVS set, which explicitly disables suid binaries, as suid binaries themselves are a common source of privilege escalations. We would not be able to use that if the supervisor itself had to be setuid-root, which means Sandstorm would be less secure.
* We would need to be very careful to write our setuid supervisor in a way that has no privilege escalations, which is not easy. I'd feel pretty bad if it turned out that installing Sandstorm actually created a new server-wide privilege escalation vulnerability by virtue of installing a setuid binary that could be exploited.

I'm happy to leave this bug open since it's a reasonable request, but I'm pretty wary of going down this path for the reasons above.

FWIW, userns bugs are only half of the privesc bugs we've [enumerated here](https://docs.sandstorm.io/en/latest/using/security-non-events/#linux-kernel). It's my opinion that it's not a good idea to rely on UID separation to run possibly-malicious code safely -- you really need a full sandbox, which should always disable userns.

**Comment by dwrensha** [comment]
@kentonv Can this be closed now that https://github.com/sandstorm-io/sandstorm/pull/2644 has landed?

**Comment by kentonv** [comment]
Indeed, this is now fixed!

**Comment by kentonv** [comment]
Note that the final solution avoided a setuid binary and instead requires that Sandstorm be started as root (which is already the typical way to run Sandstorm). This means we got to keep NO_NEW_PRIVS and did not have to worry about the setuid binary being attackable.

We managed to write the code in such a way that the userns approach is still supported, too, and shares most of the same code.

The inconsistent UID problem is still a problem, though we've observed that most apps are not affected.
